standard/expressions.md

```diff
@@ -1211,7 +1211,12 @@ The run-time processing of a function member invocation consists of the followin
   - `M` is invoked.
 - Otherwise, if the type of `E` is a value-type `V`, and `M` is declared or overridden in `V`:
   - `E` is evaluated. If this evaluation causes an exception, then no further steps are executed. For an instance constructor, this evaluation consists of allocating storage (typically from an execution stack) for the new object. In this case `E` is classified as a variable.
-  - If `E` is not classified as a variable, then a temporary local variable of `E`’s type is created and the value of `E` is assigned to that variable. `E` is then reclassified as a reference to that temporary local variable. The temporary variable is accessible as `this` within `M`, but not in any other way. Thus, only when `E` is a true variable is it possible for the caller to observe the changes that `M` makes to `this`.
+  - If `E` is not classified as a variable, or if `V` is not a readonly struct type (§16.2.2), and `E` is one of:
+    - an input parameter (§15.6.2.3), or
+    - a `readonly` field (§15.5.3), or
```

Comment:
A readonly field is writable in some context, like in a constructor, for example, and, I think, won't be subject to the copy there. Instead of trying to explain all this here, we could introduce a notion of a readonly and writable variable as a classification for expression. That would cover in parameters, readonly fields in specific context, ref readonly references, etc. Then here we can simply say: "If `E` is not classified as a **writable** variable then a temporary local variable of `E`’s type is created and the value of `E` is assigned to that variable. ..." Without all the bullet points.

Reply:
@BillWagner: I like this idea, but presumably you're not planning on doing that now? (Sounds like something to do soon after the C# 7 release though.)

Reply:
Rex has captured some of implications of introducing "writable" and "readonly" variables in #324.